### DISCUSSION TITLE
GPTR as API  + WebSocket Tutorial

### DISCUSSION THREAD

**Discussion by pedrorocha-net** [assafelovic/gpt-researcher#833] [Q&A]
I've been trying to understand how to use the gpt-researcher as a PIP package, in a FastAPI setup, to stream to a JS frontend. I saw the websocket manager and a bit of the frontend, but it's not clear to me how to recreate such behavior.

Do you know if is there any tutorial, example, anything, showing how to create that from beginning to end?

**Comment by ElishaKay** [comment]
Sup @pedrorocha-net 

Yes, this tutorial is available in the new & upcoming docs:

Here's the doc in the PR: [Querying the Backend via Websockets](https://github.com/assafelovic/gpt-researcher/blob/4e56eaad9edb8c473a5e1d1df2f3d3e3f4985834/docs/docs/gpt-researcher/gptr/querying-the-backend.md)

This example assumes that you have the GPTR Python Service running on localhost:8000

Disclaimer: The gptr-webhook.js file in the example probably needs to be refactored to support a multi-user application (hopefully will be refactored soon to support that) - but this is a good proof of concept for a single user querying the backend via a websockets connection.

May the force be with you 🙏

**Comment by pedrorocha-net** [comment]
Thanks @ElishaKay ! Curiously, there was another person also creating something around that and sharing the code yesterday: https://github.com/assafelovic/gpt-researcher/pull/832
